rust/otel-arrow-rust/src/otlp/common.rs

```diff
@@ -0,0 +1,191 @@
+    pub id: Option<&'a UInt16Array>,
+}
+
+impl ScopeArrays<'_> {
+    fn data_type() -> DataType {
+        DataType::Struct(Fields::from(vec![
+            Field::new(
+                consts::NAME,
+                DataType::Dictionary(Box::new(DataType::UInt8), Box::new(DataType::Utf8)),
+                true,
+            ),
+            Field::new(consts::VERSION, DataType::Utf8, true),
+            Field::new(consts::DROPPED_ATTRIBUTES_COUNT, DataType::UInt32, true),
+            Field::new(consts::ID, DataType::UInt16, true),
+        ]))
+    }
+
+    pub fn create_instrumentation_scope(&self, idx: usize) -> InstrumentationScope {
+        InstrumentationScope {
+            name: self.name.value_at(idx).unwrap_or_default(),
+            version: self.version.value_at_or_default(idx),
+            dropped_attributes_count: self.dropped_attributes_count.value_at_or_default(idx),
+            attributes: vec![],
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a RecordBatch> for ScopeArrays<'a> {
+    type Error = error::Error;
+
+    fn try_from(rb: &'a RecordBatch) -> Result<Self, Self::Error> {
+        let scope_array = Downcaster {
+            name: consts::SCOPE,
+            source: rb,
+            array: |rb: &'a RecordBatch| rb.column_by_name(consts::SCOPE),
+            expect_type: Self::data_type,
+        }
+        .downcast::<StructArray>()?;
+
+        let name = scope_array
+            .column_by_name(consts::NAME)
+            .map(StringArrayAccessor::try_new)
+            .transpose()?;
+
+        let version = scope_array
+            .column_by_name(consts::VERSION)
+            .map(StringArrayAccessor::try_new)
+            .transpose()?;
+
+        let dropped_attributes_count = scope_array
+            .column_by_name(consts::DROPPED_ATTRIBUTES_COUNT)
+            .map(|a| {
+                a.as_any().downcast_ref::<UInt32Array>().context(
+                    error::ColumnDataTypeMismatchSnafu {
+                        name: consts::DROPPED_ATTRIBUTES_COUNT,
+                        expect: DataType::UInt32,
+                        actual: a.data_type().clone(),
+                    },
+                )
+            })
+            .transpose()?;
+
+        let id = scope_array
+            .column_by_name(consts::ID)
+            .map(|a| {
+                a.as_any().downcast_ref::<UInt16Array>().with_context(|| {
+                    error::ColumnDataTypeMismatchSnafu {
+                        name: consts::ID,
+                        expect: DataType::UInt16,
+                        actual: a.data_type().clone(),
+                    }
+                })
+            })
+            .transpose()?;
+
+        Ok(Self {
+            name,
+            version,
+            dropped_attributes_count,
+            id,
+        })
+    }
+}
+
+struct Downcaster<S, F> {
+    name: &'static str,
+    source: S,
+    array: F,
+    expect_type: fn() -> DataType,
+}
```

Comment:
```suggestion
struct Downcaster<S, F> {
    // name is the column name, e.g. crate::schema::consts::RESOURCE
    name: &'static str,
    // source is a record batch (determines lifetime of the downcast array)
    source: S,
    // array is a function returning an optional array ref e.g., `|s| s.column_by_name(name)`
    array: F,
    // expect_type is used when downcasting fails, included in the result context.
    expect_type: fn() -> DataType,
}
```

I have been curious about this type, and now that you're moving it I'm inspired to ask. IIUC this `expect_type` is an expensive call but only called in the failure case, and I wanted to be sure.